sdk/storage/azure-storage-files-datalake/src/datalake_file_system_client.cpp

```diff
@@ -425,4 +425,71 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
         std::move(renamedDirectoryClient), std::move(result.RawResponse));
   }
 
+  ListDeletedPathsPagedResponse DataLakeFileSystemClient::ListDeletedPaths(
+      const ListDeletedPathsOptions& options,
+      const Azure::Core::Context& context) const
+  {
+    Blobs::_detail::BlobContainerClient::ListBlobContainerBlobsByHierarchyOptions
+        protocolLayerOptions;
+    protocolLayerOptions.Prefix = options.PathPrefix;
+    protocolLayerOptions.MaxResults = options.PageSizeHint;
+    protocolLayerOptions.Marker = options.ContinuationToken;
+    protocolLayerOptions.ShowOnly = Blobs::Models::ListBlobsShowOnlyType::Deleted.ToString();
+    auto result = Blobs::_detail::BlobContainerClient::ListBlobsByHierarchy(
+        *m_pipeline, m_blobContainerClient.m_blobContainerUrl, protocolLayerOptions, context);
+
+    ListDeletedPathsPagedResponse pagedResponse;
+    for (auto& item : result.Value.Items)
+    {
+      Models::PathDeletedItem pathDeletedItem;
+      if (item.Name.Encoded)
+      {
+        pathDeletedItem.Name = Core::Url::Decode(item.Name.Content);
+      }
+      else
+      {
+        pathDeletedItem.Name = std::move(item.Name.Content);
+      }
+      pathDeletedItem.DeletedOn = item.Details.DeletedOn;
+      pathDeletedItem.DeletionId = item.DeletionId;
+      pathDeletedItem.RemainingRetentionDays = item.Details.RemainingRetentionDays.Value();
+
+      pagedResponse.DeletedPaths.push_back(pathDeletedItem);
+    }
+    pagedResponse.m_operationOptions = options;
+    pagedResponse.m_fileSystemClient = std::make_shared<DataLakeFileSystemClient>(*this);
+    pagedResponse.CurrentPageToken = options.ContinuationToken.ValueOr(std::string());
+    pagedResponse.NextPageToken = result.Value.ContinuationToken;
+    pagedResponse.RawResponse = std::move(result.RawResponse);
+
+    return pagedResponse;
+  }
+
+  Azure::Response<DataLakePathClient> DataLakeFileSystemClient::Undelete(
+      const std::string& deletedPath,
+      const std::string& deletionId,
+      const Azure::Core::Context& context) const
+  {
+    std::string undeleteSource = std::string("?") + _detail::DataLakeDeletionId + "=" + deletionId;
+
+    auto blobUrl = m_blobContainerClient.m_blobContainerUrl;
+    blobUrl.AppendPath(_internal::UrlEncodePath(deletedPath));
+
+    _detail::PathClient::UndeletePathOptions options;
+    options.UndeleteSource = undeleteSource;
```

Comment:
It says on the [API reference](https://learn.microsoft.com/en-us/rest/api/storageservices/undelete-blob) `The format is blobPath?deletionid=<id>`